R-package/src/cmake/modules/FindLibR.cmake

```diff
@@ -0,0 +1,229 @@
+    # 'R' should not be used without a path -- see par. 1.6 of the manual
+    find_program(
+      LIBR_EXECUTABLE
+      NO_DEFAULT_PATH
+      HINTS "${CMAKE_CURRENT_BINARY_DIR}" "/usr/bin" "/usr/lib/"
+      NAMES R R.exe
+    )
+
+    if(LIBR_EXECUTABLE MATCHES ".*lightgbm\\.Rcheck.*")
+      message(FATAL_ERROR "If you are seeing this error, it means you are running R CMD check and R is using '${LIBR_EXECUTABLE}'. Edit src/cmake/modulesFindLibR.cmake and add your R path to HINTS near this error")
+    endif()
+  endif()
+
+  if(UNIX)
+
+    if(NOT LIBR_EXECUTABLE)
+      message(FATAL_ERROR "Unable to locate R executable.\
+        \nEither add its location to PATH or provide it through the LIBR_EXECUTABLE cmake variable")
+    endif()
+
+    # ask R for the home path
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--vanilla" "-e" "cat(R.home())"
+      OUTPUT_VARIABLE LIBR_HOME
+    )
+
+    # ask R for the include dir
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(R.home('include'))"
+      OUTPUT_VARIABLE LIBR_INCLUDE_DIRS
+    )
+
+    # ask R for the lib dir
+    execute_process(
+      COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(R.home('lib'))"
+      OUTPUT_VARIABLE LIBR_LIB_DIR
+    )
+
+  # Windows
+  else()
+
+    # ask R for R_HOME
+    if(LIBR_EXECUTABLE)
+      execute_process(
+        COMMAND ${LIBR_EXECUTABLE} "--slave" "--no-save" "-e" "cat(normalizePath(R.home(),winslash='/'))"
+        OUTPUT_VARIABLE LIBR_HOME
+      )
+    endif()
+
+    # if R executable not available, query R_HOME path from registry
+    if(NOT LIBR_HOME)
+
+      # Try to find R's location in the registry
+      # ref: https://cran.r-project.org/bin/windows/base/rw-FAQ.html#Does-R-use-the-Registry_003f
+      get_filename_component(
+        LIBR_HOME
+        "[HKEY_LOCAL_MACHINE\\SOFTWARE\\R-core\\R\\${CMAKE_R_VERSION};InstallPath]"
+        ABSOLUTE
+      )
+
+      if(NOT LIBR_HOME)
+        get_filename_component(
+          LIBR_HOME
+          "[HKEY_CURRENT_USER\\SOFTWARE\\R-core\\R\\${CMAKE_R_VERSION};InstallPath]"
+          ABSOLUTE
+        )
+      endif()
+
+      if(NOT LIBR_HOME)
+        message(FATAL_ERROR "\nUnable to locate R executable.\
+          \nEither add its location to PATH or provide it through the LIBR_EXECUTABLE cmake variable")
+      endif()
+
+    endif()
+
+    # set exe location based on R_ARCH
+    if(NOT LIBR_EXECUTABLE)
+      set(LIBR_EXECUTABLE "${LIBR_HOME}/bin/${R_ARCH}/R.exe")
+    endif()
+
+    # set other R paths based on home path
+    set(LIBR_INCLUDE_DIRS "${LIBR_HOME}/include")
+    set(LIBR_LIB_DIR "${LIBR_HOME}/bin/${R_ARCH}")
+
+    message(STATUS "LIBR_HOME [${LIBR_HOME}]")
+    message(STATUS "LIBR_EXECUTABLE [${LIBR_EXECUTABLE}]")
+    message(STATUS "LIBR_INCLUDE_DIRS [${LIBR_INCLUDE_DIRS}]")
+    message(STATUS "LIBR_LIB_DIR [${LIBR_LIB_DIR}]")
+    message(STATUS "LIBR_CORE_LIBRARY [${LIBR_CORE_LIBRARY}]")
```

Comment:
We output this info in the main CMakeLists. I don't think we should do it twice. Moreover, we haven't set `LIBR_CORE_LIBRARY` yet.

Reply:
Good idea! I think main CMakeLists is probably the right place. Because that way we're only printing after everything has been found, and the printing will also tell us if variables are being set here in the way that the main CmakeLists expects

Reply:
Agree!